src/socket/mod.rs

```diff
@@ -85,36 +81,29 @@ impl<'a, 'b> Socket<'a, 'b> {
 }
 
 /// A conversion trait for network sockets.
-///
-/// This trait is used to concisely downcast [Socket](trait.Socket.html) values to their
-/// concrete types.
-pub trait AsSocket<T> {
-    fn as_socket(&mut self) -> &mut T;
-    fn try_as_socket(&mut self) -> Option<&mut T>;
+pub trait FromSocket<'a, 'b>: SocketSession + Sized {
+    fn from_socket<'c>(&'c mut Socket<'a, 'b>) -> Option<&'c mut Self>;
```

Comment:
`downcast_socket` maybe? `from_socket` is not really indicative of what this does.